modules/renter/worker.go

```diff
@@ -0,0 +1,287 @@
+		// with.
+		contractID types.FileContractID
+
+		// If there is work on all three channels, the worker will first do all
+		// of the work in the download chan, then all of the work in the
+		// priority upload chan, and finally all of the work in the upload
+		// chan.
+		//
+		// A busy higher priority channel is able to entriely starve all of the
+		// channels with lower priority.
+		downloadChan         chan downloadWork // higher priority than all uploads
+		killChan             chan struct{}     // highest priority
+		priorityDownloadChan chan downloadWork // higher priority than downloads (used for user-initiated downloads)
+		uploadChan           chan uploadWork   // lowest priority
+
+		// recentUploadFailure documents the most recent time that an upload
+		// has failed. The repair loop ignores workers that have had an upload
+		// failure in the past two hours.
+		recentUploadFailure time.Time
+
+		// Utilities
+		renter *Renter
+	}
+)
+
+// download will perform some download work.
+func (w *worker) download(dw downloadWork) {
+	d, err := w.renter.hostContractor.Downloader(w.contractID)
+	if err != nil {
+		dw.resultChan <- finishedDownload{dw.chunkDownload, nil, err, dw.pieceIndex, w.contractID}
+		return
+	}
+	defer d.Close()
+
+	data, err := d.Sector(dw.dataRoot)
+	dw.resultChan <- finishedDownload{dw.chunkDownload, data, err, dw.pieceIndex, w.contractID}
+}
+
+// upload will perform some upload work.
+func (w *worker) upload(uw uploadWork) {
+	e, err := w.renter.hostContractor.Editor(w.contractID)
+	if err != nil {
+		w.recentUploadFailure = time.Now()
+		select {
+		case uw.resultChan <- finishedUpload{uw.chunkID, crypto.Hash{}, err, uw.pieceIndex, w.contractID}:
+		case <-w.renter.tg.StopChan():
+		}
+		return
+	}
+	defer e.Close()
+
+	root, err := e.Upload(uw.data)
+	if err != nil {
+		w.recentUploadFailure = time.Now()
+		select {
+		case uw.resultChan <- finishedUpload{uw.chunkID, root, err, uw.pieceIndex, w.contractID}:
+		case <-w.renter.tg.StopChan():
+		}
+		return
+	}
+
+	// Update the renter metadata.
+	id := w.renter.mu.Lock()
+	uw.file.mu.Lock()
+	contract, exists := uw.file.contracts[w.contractID]
+	if !exists {
+		contract = fileContract{
+			ID:          w.contractID,
+			IP:          e.Address(),
+			WindowStart: e.EndHeight(),
+		}
+	}
+	contract.Pieces = append(contract.Pieces, pieceData{
+		Chunk:      uw.chunkID.index,
+		Piece:      uw.pieceIndex,
+		MerkleRoot: root,
+	})
+	uw.file.contracts[w.contractID] = contract
+	w.renter.saveFile(uw.file)
+	uw.file.mu.Unlock()
+	w.renter.mu.Unlock(id)
+
+	select {
+	case uw.resultChan <- finishedUpload{uw.chunkID, root, err, uw.pieceIndex, w.contractID}:
+	case <-w.renter.tg.StopChan():
+	}
+}
+
+// work will perform one unit of work, exiting early if there is a kill signal
+// given before work is completed.
+func (w *worker) work() {
+	// Check for priority downloads.
+	select {
+	case d := <-w.priorityDownloadChan:
+		w.download(d)
+		return
+	default:
```

Comment:
is it possible to interrupt `w.upload`/`w.download`? Otherwise you're stuck waiting for a transfer to complete before shutdown will finish. Not terrible, but annoying.

I guess if you wanted to do it properly, you'd need to `Close` the actual connection and handle that gracefully, which is probably nontrivial.